File: RPG/v2/RPG.py
import tkinter as tk
import math

TILE = 40
ROWS, COLS = 12, 20
WIDTH, HEIGHT = COLS * TILE, ROWS * TILE

game_map = [
    [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
    [1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,1],
    [1,0,4,0,0,0,0,0,0,0,0,4,0,0,0,0,0,0,0,1],
    [1,0,0,1,1,1,0,0,0,0,0,0,1,1,1,0,0,4,0,1],
    [1,0,0,1,0,0,0,4,0,3,0,0,0,0,1,0,0,0,0,1],
    [1,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,4,0,0,1],
    [1,0,0,1,0,4,0,0,0,0,0,0,4,0,1,0,0,0,0,1],
    [1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1],
    [1,0,4,0,0,0,0,0,0,0,0,0,0,4,0,0,0,0,0,1],
    [1,0,0,0,0,0,1,1,1,1,1,1,1,0,0,0,0,4,0,1],
    [1,0,0,0,0,0,0,0,0,0,4,0,0,0,0,0,0,0,0,1],
    [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
]

class EnhancedZeldaRPG:
    def __init__(self):
        self.root = tk.Tk()
        self.root.title("Enhanced Python Zelda-Style RPG")
        self.canvas = tk.Canvas(self.root, width=WIDTH, height=HEIGHT, bg="black")
        self.canvas.pack()

        self.player_x = 400
        self.player_y = 300
        self.size = 30
        self.health = 5
        self.coins = 0
        self.enemy_x = 210
        self.enemy_y = 210
        self.enemy_health = 4
        self.keys = set()
        self.anim = 0
        self.game_over = False
        self.win = False

        self.root.bind("<KeyPress>", self.key_press)
        self.root.bind("<KeyRelease>", self.key_release)

        self.loop()
        self.root.mainloop()

    def key_press(self, e):
        self.keys.add(e.keysym.lower())
        if e.keysym == "space":
            self.attack()

    def key_release(self, e):
        self.keys.discard(e.keysym.lower())

    def draw_world(self):
        self.canvas.create_rectangle(0, 0, WIDTH, HEIGHT, fill="#1e5f28", outline="")

        for r in range(ROWS):
            for c in range(COLS):
                x, y = c * TILE, r * TILE
                val = game_map[r][c]

                if val == 1:
                    self.canvas.create_rectangle(x, y, x+TILE, y+TILE, fill="#503219", outline="#2d1b0d")
                    self.canvas.create_rectangle(x+5, y+5, x+TILE-5, y+TILE-5, fill="#785028", outline="")
                else:
                    self.canvas.create_rectangle(x, y, x+TILE, y+TILE, fill="#378737", outline="#2f7730")
                    self.canvas.create_oval(x+7, y+8, x+13, y+14, fill="#46a046", outline="")
                    self.canvas.create_oval(x+24, y+25, x+30, y+31, fill="#46a046", outline="")

                if val == 2:
                    self.draw_coin(x+20, y+20)
                elif val == 3:
                    self.draw_portal(x+20, y+20)
                elif val == 4:
                    self.draw_tree(x, y)

    def draw_coin(self, cx, cy):
        self.canvas.create_oval(cx-10, cy-10, cx+10, cy+10, fill="#ffd200", outline="#785000", width=2)
        self.canvas.create_oval(cx-5, cy-6, cx+2, cy+1, fill="#fff58c", outline="")

    def draw_portal(self, cx, cy):
        pulse = int(math.sin(self.anim * 0.1) * 5)
        self.canvas.create_oval(cx-17-pulse, cy-17-pulse, cx+17+pulse, cy+17+pulse, fill="#00ffff", outline="")
        self.canvas.create_oval(cx-13, cy-13, cx+13, cy+13, fill="#0050b4", outline="white", width=2)

    def draw_tree(self, x, y):
        self.canvas.create_rectangle(x+17, y+20, x+24, y+38, fill="#502d14", outline="")
        self.canvas.create_oval(x+7, y+5, x+33, y+31, fill="#196423", outline="")
        self.canvas.create_oval(x+13, y, x+33, y+20, fill="#28872d", outline="")

    def draw_player(self):
        x, y = self.player_x, self.player_y
        self.canvas.create_oval(x-3, y+24, x+36, y+36, fill="#111111", outline="")
        self.canvas.create_rectangle(x, y+8, x+self.size, y+self.size+4, fill="#285ae6", outline="#102e91")
        self.canvas.create_oval(x+5, y, x+25, y+20, fill="#f5cda0", outline="#8a5a35")
        self.canvas.create_polygon(x+6, y+4, x+15, y-13, x+24, y+4, fill="#143ca0")
        self.canvas.create_oval(x+9, y+8, x+13, y+12, fill="white")
        self.canvas.create_oval(x+18, y+8, x+22, y+12, fill="white")
        self.canvas.create_rectangle(x+28, y+14, x+46, y+19, fill="#d0d0d0", outline="white")

    def draw_enemy(self):
        if self.enemy_health <= 0:
            return
        x, y = self.enemy_x, self.enemy_y
        self.canvas.create_oval(x-3, y+24, x+36, y+36, fill="#111111", outline="")
        self.canvas.create_oval(x, y, x+32, y+32, fill="#aa141e", outline="#500000")
        self.canvas.create_oval(x+6, y+9, x+12, y+15, fill="#500000")
        self.canvas.create_oval(x+20, y+9, x+26, y+15, fill="#500000")
        self.canvas.create_text(x+16, y-8, text=f"HP {self.enemy_health}", fill="white", font=("Arial", 10, "bold"))

    def draw_hud(self):
        self.canvas.create_rectangle(10, 10, 465, 48, fill="#000000", outline="#444444")
        for i in range(5):
            color = "red" if i < self.health else "#333333"
            self.canvas.create_oval(25+i*28, 18, 45+i*28, 38, fill=color, outline="")
        self.canvas.create_text(215, 30, text=f"Coins: {self.coins}", fill="white", font=("Arial", 16, "bold"))
        self.canvas.create_text(360, 30, text="Move: WASD/Arrows  Attack: SPACE", fill="white", font=("Arial", 12, "bold"))

    def draw(self):
        self.canvas.delete("all")
        self.draw_world()
        self.draw_enemy()
        self.draw_player()
        self.draw_hud()

        if self.game_over or self.win:
            self.canvas.create_rectangle(0, 250, WIDTH, 350, fill="#000000")
            text = "GAME OVER" if self.game_over else "YOU FOUND THE PORTAL!"
            color = "red" if self.game_over else "cyan"
            self.canvas.create_text(WIDTH//2, 310, text=text, fill=color, font=("Arial", 36, "bold"))

    def hits_wall(self, x, y):
        corners = [(x, y), (x+self.size, y), (x, y+self.size), (x+self.size, y+self.size)]
        for px, py in corners:
            r, c = py // TILE, px // TILE
            if r < 0 or r >= ROWS or c < 0 or c >= COLS:
                return True
            if game_map[r][c] == 1:
                return True
        return False

    def move(self):
        nx, ny = self.player_x, self.player_y
        if "w" in self.keys or "up" in self.keys: ny -= 5
        if "s" in self.keys or "down" in self.keys: ny += 5
        if "a" in self.keys or "left" in self.keys: nx -= 5
        if "d" in self.keys or "right" in self.keys: nx += 5
        if not self.hits_wall(nx, ny):
            self.player_x, self.player_y = nx, ny

    def intersects(self, a, b):
        ax, ay, aw, ah = a
        bx, by, bw, bh = b
        return ax < bx+bw and ax+aw > bx and ay < by+bh and ay+ah > by

    def check_tiles_enemy(self):
        r = (self.player_y + self.size//2) // TILE
        c = (self.player_x + self.size//2) // TILE
        if game_map[r][c] == 2:
            self.coins += 1
            game_map[r][c] = 0
        if game_map[r][c] == 3:
            self.win = True

        if self.enemy_health > 0 and self.intersects(
            (self.player_x, self.player_y, self.size, self.size),
            (self.enemy_x, self.enemy_y, 32, 32)
        ):
            self.health -= 1
            self.player_x, self.player_y = 400, 300
            if self.health <= 0:
                self.game_over = True

    def attack(self):
        if self.enemy_health > 0 and self.intersects(
            (self.player_x-25, self.player_y-25, self.size+50, self.size+50),
            (self.enemy_x, self.enemy_y, 32, 32)
        ):
            self.enemy_health -= 1

    def loop(self):
        self.anim += 1
        if not self.game_over and not self.win:
            self.move()
            self.check_tiles_enemy()
        self.draw()
        self.root.after(16, self.loop)

EnhancedZeldaRPG()

File: RPG/v2/RPG.java
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

public class RPG extends JFrame {
    public RPG() {
        add(new RPGPanel());
        setTitle("Mini RPG");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);
        pack();
        setLocationRelativeTo(null);
        setVisible(true);
    }

    public static void main(String[] args) {
        new RPG();
    }
}

class RPGPanel extends JPanel implements ActionListener, KeyListener {
    private final int TILE = 40;
    private final int WIDTH = 800;
    private final int HEIGHT = 600;

    private final javax.swing.Timer timer = new javax.swing.Timer(16, this);
    private boolean up, down, left, right;
    private int playerX = 400, playerY = 300;
    private int playerSize = 30;
    private int health = 5;
    private int coins = 0;
    private int enemyX = 210, enemyY = 210;
    private int enemyHealth = 4;
    private int anim = 0;
    private boolean gameOver = false, win = false;

    private int[][] map = {
        {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1},
        {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,1},
        {1,0,4,0,0,0,0,0,0,0,0,4,0,0,0,0,0,0,0,1},
        {1,0,0,1,1,1,0,0,0,0,0,0,1,1,1,0,0,4,0,1},
        {1,0,0,1,0,0,0,4,0,3,0,0,0,0,1,0,0,0,0,1},
        {1,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,4,0,0,1},
        {1,0,0,1,0,4,0,0,0,0,0,0,4,0,1,0,0,0,0,1},
        {1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1},
        {1,0,4,0,0,0,0,0,0,0,0,0,0,4,0,0,0,0,0,1},
        {1,0,0,0,0,0,1,1,1,1,1,1,1,0,0,0,0,4,0,1},
        {1,0,0,0,0,0,0,0,0,0,4,0,0,0,0,0,0,0,0,1},
        {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1}
    };

    public RPGPanel() {
        setPreferredSize(new Dimension(WIDTH, HEIGHT));
        setFocusable(true);
        addKeyListener(this);
        timer.start();
    }

    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2 = (Graphics2D) g;
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        drawWorld(g2);
        drawEnemy(g2);
        drawPlayer(g2);
        drawHUD(g2);

        if (gameOver) drawCenter(g2, "GAME OVER", Color.RED);
        if (win) drawCenter(g2, "YOU FOUND THE PORTAL!", Color.CYAN);
    }

    private void drawWorld(Graphics2D g2) {
        GradientPaint bg = new GradientPaint(0, 0, new Color(30, 95, 40), 0, HEIGHT, new Color(12, 55, 30));
        g2.setPaint(bg);
        g2.fillRect(0, 0, WIDTH, HEIGHT);

        for (int r = 0; r < map.length; r++) {
            for (int c = 0; c < map[0].length; c++) {
                int x = c * TILE, y = r * TILE;

                if (map[r][c] == 1) {
                    g2.setColor(new Color(80, 50, 25));
                    g2.fillRoundRect(x, y, TILE, TILE, 10, 10);
                    g2.setColor(new Color(120, 80, 40));
                    g2.fillRoundRect(x + 5, y + 5, TILE - 10, TILE - 10, 8, 8);
                } else {
                    g2.setColor(new Color(55, 135, 55));
                    g2.fillRect(x, y, TILE, TILE);
                    g2.setColor(new Color(70, 160, 70, 120));
                    g2.fillOval(x + 6, y + 8, 6, 6);
                    g2.fillOval(x + 24, y + 25, 5, 5);
                }

                if (map[r][c] == 2) drawCoin(g2, x + 20, y + 20);
                if (map[r][c] == 3) drawPortal(g2, x + 20, y + 20);
                if (map[r][c] == 4) drawTree(g2, x, y);
            }
        }
    }

    private void drawCoin(Graphics2D g2, int cx, int cy) {
        g2.setColor(new Color(255, 210, 0));
        g2.fillOval(cx - 10, cy - 10, 20, 20);
        g2.setColor(new Color(255, 245, 140));
        g2.fillOval(cx - 5, cy - 6, 7, 7);
        g2.setColor(new Color(120, 80, 0));
        g2.drawOval(cx - 10, cy - 10, 20, 20);
    }

    private void drawPortal(Graphics2D g2, int cx, int cy) {
        int pulse = (int)(Math.sin(anim * 0.1) * 5);
        g2.setColor(new Color(0, 255, 255, 120));
        g2.fillOval(cx - 17 - pulse, cy - 17 - pulse, 34 + pulse * 2, 34 + pulse * 2);
        g2.setColor(new Color(0, 80, 180));
        g2.fillOval(cx - 13, cy - 13, 26, 26);
        g2.setColor(Color.WHITE);
        g2.drawArc(cx - 10, cy - 10, 20, 20, anim * 8, 240);
    }

    private void drawTree(Graphics2D g2, int x, int y) {
        g2.setColor(new Color(80, 45, 20));
        g2.fillRect(x + 17, y + 20, 7, 18);
        g2.setColor(new Color(25, 100, 35));
        g2.fillOval(x + 7, y + 5, 26, 26);
        g2.setColor(new Color(40, 135, 45));
        g2.fillOval(x + 13, y, 20, 20);
    }

    private void drawPlayer(Graphics2D g2) {
        g2.setColor(new Color(20, 20, 20, 90));
        g2.fillOval(playerX - 3, playerY + 24, 36, 12);

        g2.setColor(new Color(40, 90, 230));
        g2.fillRoundRect(playerX, playerY + 8, playerSize, playerSize - 4, 12, 12);

        g2.setColor(new Color(245, 205, 160));
        g2.fillOval(playerX + 5, playerY, 20, 20);

        g2.setColor(new Color(20, 60, 160));
        g2.fillPolygon(new int[]{playerX + 6, playerX + 15, playerX + 24}, new int[]{playerY + 4, playerY - 13, playerY + 4}, 3);

        g2.setColor(Color.WHITE);
        g2.fillOval(playerX + 9, playerY + 8, 4, 4);
        g2.fillOval(playerX + 18, playerY + 8, 4, 4);

        g2.setColor(Color.LIGHT_GRAY);
        g2.fillRect(playerX + 28, playerY + 14, 18, 5);
        g2.setColor(Color.WHITE);
        g2.drawRect(playerX + 28, playerY + 14, 18, 5);
    }

    private void drawEnemy(Graphics2D g2) {
        if (enemyHealth <= 0) return;

        g2.setColor(new Color(20, 20, 20, 90));
        g2.fillOval(enemyX - 3, enemyY + 24, 36, 12);
        g2.setColor(new Color(170, 20, 30));
        g2.fillOval(enemyX, enemyY, 32, 32);
        g2.setColor(new Color(80, 0, 0));
        g2.fillOval(enemyX + 6, enemyY + 9, 6, 6);
        g2.fillOval(enemyX + 20, enemyY + 9, 6, 6);
        g2.setColor(Color.WHITE);
        g2.drawString("HP " + enemyHealth, enemyX - 3, enemyY - 5);
    }

    private void drawHUD(Graphics2D g2) {
        g2.setColor(new Color(0, 0, 0, 160));
        g2.fillRoundRect(10, 10, 455, 38, 15, 15);

        for (int i = 0; i < 5; i++) {
            g2.setColor(i < health ? Color.RED : Color.DARK_GRAY);
            g2.fillOval(25 + i * 28, 18, 20, 20);
        }

        g2.setColor(Color.WHITE);
        g2.setFont(new Font("Arial", Font.BOLD, 17));
        g2.drawString("Coins: " + coins, 180, 35);
        g2.drawString("Move: WASD/Arrows   Attack: SPACE", 285, 35);
    }

    private void drawCenter(Graphics2D g2, String text, Color color) {
        g2.setColor(new Color(0, 0, 0, 170));
        g2.fillRect(0, 250, WIDTH, 100);
        g2.setColor(color);
        g2.setFont(new Font("Arial", Font.BOLD, 42));
        FontMetrics fm = g2.getFontMetrics();
        g2.drawString(text, (WIDTH - fm.stringWidth(text)) / 2, 315);
    }

    public void actionPerformed(ActionEvent e) {
        anim++;
        if (!gameOver && !win) {
            move();
            checkTiles();
            checkEnemy();
        }
        repaint();
    }

    private void move() {
        int nx = playerX, ny = playerY;
        if (up) ny -= 5;
        if (down) ny += 5;
        if (left) nx -= 5;
        if (right) nx += 5;

        if (!hitsWall(nx, ny)) {
            playerX = nx;
            playerY = ny;
        }
    }

    private boolean hitsWall(int x, int y) {
        Rectangle player = new Rectangle(x, y, playerSize, playerSize);
        for (int r = 0; r < map.length; r++) {
            for (int c = 0; c < map[0].length; c++) {
                if (map[r][c] == 1) {
                    Rectangle wall = new Rectangle(c * TILE, r * TILE, TILE, TILE);
                    if (player.intersects(wall)) return true;
                }
            }
        }
        return false;
    }

    private void checkTiles() {
        int row = (playerY + playerSize / 2) / TILE;
        int col = (playerX + playerSize / 2) / TILE;

        if (map[row][col] == 2) {
            coins++;
            map[row][col] = 0;
        }
        if (map[row][col] == 3) win = true;
    }

    private void checkEnemy() {
        if (enemyHealth <= 0) return;
        Rectangle player = new Rectangle(playerX, playerY, playerSize, playerSize);
        Rectangle enemy = new Rectangle(enemyX, enemyY, 32, 32);
        if (player.intersects(enemy)) {
            health--;
            playerX = 400;
            playerY = 300;
            if (health <= 0) gameOver = true;
        }
    }

    private void attack() {
        if (enemyHealth <= 0) return;
        Rectangle attack = new Rectangle(playerX - 25, playerY - 25, playerSize + 50, playerSize + 50);
        Rectangle enemy = new Rectangle(enemyX, enemyY, 32, 32);
        if (attack.intersects(enemy)) enemyHealth--;
    }

    public void keyPressed(KeyEvent e) {
        int k = e.getKeyCode();
        if (k == KeyEvent.VK_W || k == KeyEvent.VK_UP) up = true;
        if (k == KeyEvent.VK_S || k == KeyEvent.VK_DOWN) down = true;
        if (k == KeyEvent.VK_A || k == KeyEvent.VK_LEFT) left = true;
        if (k == KeyEvent.VK_D || k == KeyEvent.VK_RIGHT) right = true;
        if (k == KeyEvent.VK_SPACE) attack();
    }

    public void keyReleased(KeyEvent e) {
        int k = e.getKeyCode();
        if (k == KeyEvent.VK_W || k == KeyEvent.VK_UP) up = false;
        if (k == KeyEvent.VK_S || k == KeyEvent.VK_DOWN) down = false;
        if (k == KeyEvent.VK_A || k == KeyEvent.VK_LEFT) left = false;
        if (k == KeyEvent.VK_D || k == KeyEvent.VK_RIGHT) right = false;
    }

    public void keyTyped(KeyEvent e) {}
}
